### ISSUE TITLE
Possibilty to clamp hidden layer activation?

### ISSUE THREAD

**Issue by tuanpham96** [NN4Neurosim/nn4n#3]
Been trying this out and it's been great so far! 

One question regarding bounds of hidden layer activations. Sometimes, the activations of the hidden layer units explode to very large values (e.g. more than 1E10), and it just basically an exponential increase over time points. 

I tried using FR and hidden loss in RNNLoss but it can still happen, but not very often, especially if I crank up the `lambda` for these losses. An alternative strategy would be to clamp the activation of the hidden unit after ReLu.

Is it recommended and possible within the package (maybe some form of bounded ReLu)?

Thanks!

EDIT: So I figured I could try using using `activation` as Hardtanh to define the bounds. However it seems that this currently is not allowed as the inputs are limited?

https://github.com/NN4Neurosim/nn4n/blob/103f7e2d688ad3f524009868277c2115eb913b89/nn4n/utils/helper_functions.py#L41-L59

**Comment by zhaozewang** [comment]
Hi @tuanpham96, thank you for your feedback.

One simple fix is to add the additional activation to the `def get_activation(act): `, you can do so by cloning from the source, modifying this function, cd to the project root, and installing the package via `pip install -e .`.

Another way is to check out the branch `block_recurrent`, which allows for a much more flexible definition of RNNs and their activation functions. But this branch hasn't been well-documented at the moment. Here's an example code of defining the RNN with code on this new branch:

```python
import nn4n
import torch


model_cfg = {
    "input_dim": 10,
    "hidden_dim": 100,
    "output_dim": 10,
    "alpha": 0.1,
    "learn_alpha": False,
    "preact_noise": False,
    "postact_noise": False
}
rnn = nn4n.nn.RNN(
    recurrent_layers=[
        nn4n.nn.RecurrentLayer(
            projection_layer=nn4n.nn.LinearLayer(
                input_dim=model_cfg["input_dim"], 
                output_dim=model_cfg["hidden_dim"]
            ),
            leaky_layer=nn4n.nn.LeakyLinearLayer(
                linear_layer=nn4n.nn.LinearLayer(
                    input_dim=model_cfg["hidden_dim"], 
                    output_dim=model_cfg["hidden_dim"]
                ),
                activation=torch.nn.ReLU(),
                alpha=model_cfg["alpha"],
                learn_alpha=model_cfg["learn_alpha"],
                preact_noise=model_cfg["preact_noise"],
                postact_noise=model_cfg["postact_noise"]
            )
        )
    ],
    readout_layer=nn4n.nn.LinearLayer(
        input_dim=model_cfg["hidden_dim"], 
        output_dim=model_cfg["output_dim"]
    )
)
```
You may change the activation to any customized function as long as it follows the standard torch-style activation function definition. 

The `nn4n.nn.RNN` takes a list of `RecurrentLayer` objects, so that one can define an RNN with multiple hidden layers. Each `RecurrentLayer` object optionally takes a projection layer (which corresponds to the input projection in v1.2.0) and a leaky layer (which corresponds to the `hidden_layer` in v1.2.0). And the RNN optionally takes a `readout_layer` so that one can directly readout the hidden state or let the hidden state to be projected out by a learned projection matrix.

**Comment by tuanpham96** [comment]
sorry for the late reply. thanks for the suggestions! I was able to just set the activation function without changing the internal code like this.

```python
rnn = CTRNN(...)
rnn.recurrent_layer.activation = torch.nn.Hardtanh(
     min_val=0.0,  max_val=10.0
)
```

I'll keep an eye out on that branch when I need more granular customizations of the model. thanks for pointing me towards it! I'll close this now.
